### PR TITLE
fix(agent-cli-wrapper): suppress ScheduleWakeup turn to prevent jiradozer premature exit

### PR DESCRIPTION
**Pull request by mzhaom**
## Summary

- Jiradozer was exiting prematurely when the agent called `ScheduleWakeup` (the `/loop` dynamic pacing tool). The session wrapper treated the turn as complete and `defer session.Stop()` killed the CLI before the wakeup fired.
- Added `wakeupSuppressionState` in `session.go`: `handleResult()` detects `ScheduleWakeup` via `turn.hasScheduleWakeup()`, accumulates token usage, and suppresses `TurnCompleteEvent` until the continuation turn arrives (or a safety timer fires at delay+60s).
- Adds `scheduleWakeupDelaySeconds()` helper to `turn.go` for safety timer sizing.

## Test plan

- [ ] `TestScheduleWakeup_TurnSuppressedWhenScheduleWakeupPresent` — suppression fires, no premature `TurnCompleteEvent`
- [ ] `TestScheduleWakeup_ContinuationCompletesTurn` — continuation releases suppression with original turn number and accumulated usage
- [ ] `TestScheduleWakeup_SafetyTimerCompletesTurn` — safety timer fires and completes turn if no continuation arrives
- [ ] `TestScheduleWakeup_ErrorTurnNotSuppressed` — error results bypass suppression and complete immediately
- [ ] `bazel test //agent-cli-wrapper/claude/... //jiradozer/... //multiagent/agent/...` — all pass
- [ ] `scripts/lint.sh` — passes

---

> [!NOTE]
> **Medium Risk**
> Touches core session turn-completion/timer logic and adds new cross-turn suppression state, which is concurrency-sensitive and could impact event ordering or completion semantics if incorrect.
> 
> **Overview**
> Fixes premature session completion when the agent uses `ScheduleWakeup` by **suppressing turn completion until the CLI’s delayed continuation arrives** (or a safety timer fires), while accumulating usage/cost across the suppressed chain and completing under the original turn number.
> 
> This adds wakeup-specific suppression state/timer handling, adjusts `finalizeTurn` ordering and `CollectResponse` to prefer recorded completed results (so continuation content is returned), and introduces helpers (`TurnUsage.Add`, turn `ScheduleWakeup` detection/delay extraction) plus extensive unit tests covering continuation, budget/limit, timer, and double-completion edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c4bf7fbbcf57fad7cff3e6adf5e89c730379df58. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: agent-cli-wrapper/claude/session.go [hunk 1]
```diff
@@ -1071,7 +1102,21 @@ func (s *Session) handleResult(msg protocol.ResultMessage) {
...
+	// not be emitted — return early.
+	if s.wakeupState.timer != nil {
+		s.wakeupState.timer.Stop()
+		s.wakeupState.timer = nil
+	}
+	wakeupTimerFired := s.wakeupState.timerFired
+	s.wakeupState.timerFired = false
+	wakeupSuppressed := s.wakeupState.active
+	// Don't clear wakeupState.active yet — we may need to re-suppress below
+	// if this continuation turn also contains ScheduleWakeup.
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/claude/session.go]: Fixed in 6aa615b. wakeupState.active is now cleared under the same lock where timerFired is read, mirroring bgState. Added TestScheduleWakeup_WakeupStateClearedOnNewTurn and strengthened the suppression test to drain the full event window.

File: agent-cli-wrapper/claude/session.go [hunk 2]
```diff
@@ -1272,19 +1335,86 @@ func (s *Session) handleResult(msg protocol.ResultMessage) {
...
+	shouldSuppressWakeup := !msg.IsError && turn.hasScheduleWakeup()
+	if shouldSuppressWakeup {
+		s.mu.Lock()
+		s.cumulativeCostUSD += msg.TotalCostUSD
+		// Accumulate usage from this suppressed turn.
+		s.wakeupState.accumulatedUsage.InputTokens += msg.Usage.InputTokens
+		s.wakeupState.accumulatedUsage.OutputTokens += msg.Usage.OutputTokens
+		s.wakeupState.accumulatedUsage.CacheCreationTokens += msg.Usage.CacheCreationInputTokens
+		s.wakeupState.accumulatedUsage.CacheReadTokens += msg.Usage.CacheReadInputTokens
+		s.wakeupState.accumulatedUsage.CostUSD += msg.TotalCostUSD
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/claude/session.go]: Fixed in 6aa615b. The re-suppression path now re-adds accUsage (drained from prior suppressed turns) back into wakeupState.accumulatedUsage. Added TestScheduleWakeup_ChainedSuppressionAccumulatesUsage covering three chained turns and verifying tokens/cost accumulate.

File: agent-cli-wrapper/claude/session.go [hunk 3]
```diff
@@ -1272,19 +1335,86 @@ func (s *Session) handleResult(msg protocol.ResultMessage) {
...
-	if !costAccounted {
+	if !costAccounted && !shouldSuppressWakeup {
 		s.cumulativeCostUSD += msg.TotalCostUSD
 	}
 	totalCost := s.cumulativeCostUSD
 	s.mu.Unlock()
 
 	if result.Error == nil {
-		if s.config.MaxTurns > 0 && turnNumber >= s.config.MaxTurns {
+		if s.config.MaxTurns > 0 && resultTurnNumber >= s.config.MaxTurns {
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/claude/session.go]: Fixed in 6aa615b. MaxTurns now enforces against the actual turnNumber, not resultTurnNumber. Added a code comment explaining why the two differ (resultTurnNumber is the suppressed number used to unblock waiters; the real assistant-turn count is what MaxTurns must bound).

File: agent-cli-wrapper/claude/session.go [hunk 4]
```diff
@@ -1086,14 +1149,27 @@ func (s *Session) handleResult(msg protocol.ResultMessage) {
...
 
+	// When wakeup suppression was active, waiters are blocked on the original
+	// suppressed turn number. Use that number so CompleteTurn unblocks them.
+	// (wakeupState.active was already cleared above under the initial lock.)
+	resultTurnNumber := turnNumber
+	if wakeupSuppressed {
+		s.mu.Lock()
+		resultTurnNumber = s.wakeupState.suppressedTurnNumber
+		s.mu.Unlock()
+	}
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/claude/session.go:1172]: Fixed. suppressedTurnNumber is now snapshotted under the initial lock (c061dbf — that commit was already on the PR before this review, but the line numbers cited are from the prior commit).

File: agent-cli-wrapper/claude/session.go [hunk 5]
```diff
@@ -1272,18 +1351,89 @@ func (s *Session) handleResult(msg protocol.ResultMessage) {
...
+	// Skip suppression when:
+	//   - the result carries an error (propagate immediately)
+	//   - MaxTurns has been reached (let the normal path surface it)
+	//   - the turn does not contain a ScheduleWakeup tool_use
+	// Don't suppress if we've already hit MaxTurns. Suppression would
+	// re-arm the session for another continuation turn, letting chained
+	// wakeups run past the configured limit. Fall through so the MaxTurns
+	// guard below can convert this result into ErrMaxTurnsExceeded.
+	maxTurnsReached := s.config.MaxTurns > 0 && turnNumber >= s.config.MaxTurns
+	shouldSuppressWakeup := !msg.IsError && !maxTurnsReached && turn.hasScheduleWakeup()
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/claude/session.go]: Fixed in c4bf7fb. Added the missing !wakeupSuppressed guard in shouldSuppressWakeup, mirroring the bgState !wasSuppressed pattern. Also rewrote tests that manually called StartTurn for continuations (which doesn't happen in production) — new TestScheduleWakeup_ContinuationDoesNotReSuppress covers the specific regression.
